### DISCUSSION TITLE
Solution for flows that span over two adjacent pcaps?

### DISCUSSION THREAD

**Discussion by sunny-r** [nfstream/nfstream#50] [Ideas]
tcpdump/libpcap only allows 2GB packet capture files. Presumably, you could end up with a TCP flow that spans two different 2GB files. Processing these with nfstream would produce two truncated flows instead of one, I think? 

Can it be possible to use nfstream to do a live capture off an interface, which produces flow objects with NFStreamer, but also writes pcap files to disk when they hit a certain size (I need these for a different application)?

The live capture should be able to receive a tcpdump filter, e.g.  'tcp and port not 51413'

**Comment by drnpkr** [comment]
Hey there!

I do not think overloading NFStream with writing into PCAP while organizing packets into flow records is a good idea. It can negatively impact its performance.

Did you try using n2disk? https://www.ntop.org/products/traffic-recording-replay/n2disk/

This is probably what you are looking for. I have managed to capture PCAP files with size over 10 GB. Although, after you reach your memory cap, the speed of your disk (I/O performance) becomes a bottleneck.

**Reply by sunny-r** [reply]
Hi, thanks for getting back to me.

That's fair enough, I can see that. 

n2disk looks like it could be worth a shot, but AFAIK it needs a paid licence, and I'm just a enthusiast doing this stuff at home for fun! I probably won't pay for enterprise software

**Reply by drnpkr** [reply]
Indeed, licensing can be a challenge. However, while I have no recollection on the specific value, but n2disk is fully functional for a specific time (10 minutes?) without a license.

Alternatively, you can change the source code of tcpdump and increase the size of the max file. You can start here: https://serverfault.com/questions/647697/tcpdump-maximum-split-file-size

**Reply by sunny-r** [reply]
Picking apart C++... :(

I've seen the 'PcapPlusPlus' library, but building a pcap generator to increase 2GB captures to ~10GB is a lot of work without much gain. 

No worries!

**Reply by aouinizied** [reply]
@sunny-r @drnpkr Another proposal would be as follows:

* We can extend NFStream to take into account a list of pcaps and add a parameter like `merge_mode`.
* The idea is that when this parameter is set to False. NFstream process pcap one by one. If it is set to True, NFStream processes the first and then continues to the second without flushing the cache and so on. The capture termination will be based only on the last file end of the file signal.

This will allow us to fit your need without adding packet writing stuff. What do you think guys?

Zied

**Reply by aouinizied** [reply]
Note that the ordering of the pcaps is out of the scope and the list of pcap files must be ordered by the user before passing it as input for the Streamer.

**Reply by sunny-r** [reply]
Hi Zied, that sounds interesting, and more practical too.

I'm planning to have tcpdump running on an interface continually; it will keep yielding 2GB captures into a directory.

I guess I could have nfstream run as a 'batch' processor - to pick up a bunch of ordered pcaps, produce the flows, and then wait to pick up the next batch. Still a chance to truncate some ("inter-batch" flows), but much reduced. A con of this approach is latency in processing, rather than getting near real time flows. But I'll definitely take this, sounds good.

Have you thought about having nfstream run as a daemon process? It could watch a directory for pcaps, and throws out flows when they end and caches them between pcaps, say, up to a user-specified time-period, or maximum number of pcaps before flushing from cache? Just a thought!

**Reply by aouinizied** [reply]
Another option is to run nfstream directly on live mode and stores the flows without storing the pcaps, in that case, all your worries about truncated flows are solved. Is there a specific need that makes you want to store the pcap?

**Reply by drnpkr** [reply]
I second this.

However, I think, the major problem here is the max file size limit of tcpdump that we now try to address in a layer higher, i.e. NFStream.

@sunny-r could you please provide a specific use case where this functionality might be required by other users too? Perhaps I miss something and this feature could indeed be valuable.

**Reply by aouinizied** [reply]
@drnpkr This can be used to process MAWI dataset, or any other big dataset. Administrators, even without the limit of file size, tends to split the capture into multiple files for storage and safety purposes. For example, we could process one day of MAWI capture without any flow truncation. What do you think?

**Reply by aouinizied** [reply]
Another must-have feature is to read even compressed files. In this way we can process all files without wasting decompression required storage and without flow truncation.

**Reply by drnpkr** [reply]
@aouinizied or CAIDA. Well, this is something that is definitely not supported by related work (as far as I know) and could significantly increase the practicability of NFStream. I see it now.

**Reply by sunny-r** [reply]
So my use-cases will be less important than proper network administrators, but here they are:
1) confidence and ability to error-check: I'll want to compare what I'm seeing at the packet-level (in my pcap) with the corresponding flow; I'll still have to delete the pcaps, but I can hold onto recent ones to do this comparison
2) I'm doing more detailed 'ndpi' extraction than is possible in nfstream - e.g. lots of fields from TLS certificates, DNS responses, which can be long (CNAMEs, etc.); you can't fit this into 2 or 3 fields. I've actually got this mostly done, though it's all in python, and it works off pcaps (which is why i want them in addition to nfstream)
3) the OCD in me doesn't want to lose any flows!

**Reply by aouinizied** [reply]
@sunny-r I understand your thought. What I can say is that:

* **1** We did a big part of it as when we started NFStream it was mainly due to the lack of reliability of existing measurement tools (not the same results, etc.) so we spent a lot of time checking the results and comparing pcaps to flows manually. As there no bug-free software, I understand that you need that but you can reduce it to one random file check from time to time.
* **2** Within NFStream internal structures, there is all nDPI structures exposed. So, do not hesitate to add your fields to NFStream. It's as simple as a copy from a struct field to a Python object.
* **3** And you are right! :D

**Reply by sunny-r** [reply]
Ok, I don't doubt you guys have been careful, but picking through a few examples also helps me understand what's going on, but like you said, it's only the odd random check.

I'd like to have a go at 2) - using built-in nDPI to extract fields of interest. Can you give me an example to help me get started, I've not done this before. Say we have a DNS response with two Answers as CNAMEs, how would the hosts and IPs be exposed? I'm not really sure how to go from ndpi/structs to python objects.

Thanks!

**Reply by sunny-r** [reply]
Hi,

I'm struggling to expose the pre-extracted nDPI fields in a flow; I've started writing an NFPlugin, and found an object that I've tried to pull out for some analysis:

```
    def on_expire(self, flow):
        flow.udps.ndpi_flow = flow._C.ndpi_flow
```

And this is the error I see:

```
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/queues.py", line 234, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/usr/lib/python3.6/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: can't pickle _cffi_backend._CDataBase objects
```

Can I get some help on how to extract fields already identified by the nDPI engine?

Thanks

Sunny

**Reply by sunny-r** [reply]
bump; @aouinizied - how can i expose more of the ndpi fields that are extracted? Do I need to use NFPlugin?

**Reply by aouinizied** [reply]
@sunny-r It will depends on your needs. There is mainly two ways to do it.

* The first one is to extend the source code of NFStream to handle additional fields from nDPI. 
* Using NFPlugin as you started.

Let's focus on 2 as it is the simplest way (less efficient however).

As you can see ndpi_flow field is under protected attribute of NFlow and this is because user are not supposed to play with it (specially overwriting it).
Now if you want to read it, this can be achieved, however, you must keep in mind some points about how NFStream works internally and some general knowledge about how C structures are handled within Python:

* Whatever you wan to create as a plugin field, it must be a Python object. This is due to the fact that NFStream several meters processes that communicate with the parent process using sockets and thus, everything is pickled at some point. This explain why you have this error, you assign to the attribute ndpi_flow structure which is a C structure and pickle don't know how to serialize it.
* You must keep in mind that internally (for memory saving purposes), ndpi references (flow_struct_src_id_struct and dst_id struct) are freed as soon as the detection is completed. Consequently, if you access this structure when it is freed, you will have issues.
* So, you can access it mainly on_init and on_update before detection is completed. Note that with such approach you will not be called on the packet when the detection is considered as completed (internal logic of NFStream that you can check in engine.c)

Please find in the following a working example on how to extract host_server_name for nDPI flow structure within a plugin.

```python
from nfstream import NFStreamer, NFPlugin
import cffi

class NDPIExtractor(NFPlugin):
    def on_init(self, packet, flow):
        flow.udps.my_field = None
        if flow._C.detection_completed == 0:  # This avoid attempt to read a freed ndpi_flow structure
            flow.udps.my_field = self.ffi.string(
                flow._C.ndpi_flow.protos.tls_quic_stun.tls_quic.client_requested_server_name
            ).decode('utf-8')

    def on_update(self, packet, flow):
        if flow._C.detection_completed == 0:  # This avoid attempt to read a freed ndpi_flow structure
            flow.udps.my_field = self.ffi.string(
                flow._C.ndpi_flow.protos.tls_quic_stun.tls_quic.client_requested_server_name
            ).decode('utf-8')

    # Not using on_expire to avoid manipulating ndpi_flow as it is already freed


flow_streamer = NFStreamer(source="tests/facebook.pcap",  udps=NDPIExtractor(ffi=cffi.FFI()))
for flow in flow_streamer:
   print(flow)
```

As in this example the targeted field is a char [fixed_size], some casting through cffi is needed to convert it to a string. This is not needed if you will work on integer fields of nDPI structures.

Hope this helps!

Zied
